// @enact/moonstone/EditableIntegerPicker/EditableIntegerPickerDecorator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.EditableIntegerPickerDecorator = void 0;
var _keymap = require("@enact/core/keymap");
var _handle = require("@enact/core/handle");
var _hoc = _interopRequireDefault(require("@enact/core/hoc"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _react = require("react");
var _reactDom = _interopRequireDefault(require("react-dom"));
var _spotlight = _interopRequireDefault(require("@enact/spotlight"));
var _Pause = _interopRequireDefault(require("@enact/spotlight/Pause"));
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : String(i); }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); Object.defineProperty(subClass, "prototype", { writable: false }); if (superClass) _setPrototypeOf(subClass, superClass); }
function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
function _possibleConstructorReturn(self, call) { if (call && (typeof call === "object" || typeof call === "function")) { return call; } else if (call !== void 0) { throw new TypeError("Derived constructors may only return object or undefined"); } return _assertThisInitialized(self); }
function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }
(0, _keymap.addAll)({
  minus: [109, 189],
  numSet: [48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105],
  plus: 107
});

/**
 *{@link moonstone/EditableIntegerPicker.EditableIntegerPickerDecorator} is a higher-order component
 * which handles various functionalities of {@link moonstone/EditableIntegerPicker.EditableIntegerPicker}.
 *
 * @class EditableIntegerPickerDecorator
 * @memberof moonstone/EditableIntegerPicker
 * @hoc
 * @private
 */
var EditableIntegerPickerDecorator = exports.EditableIntegerPickerDecorator = (0, _hoc["default"])(function (config, Wrapped) {
  var _class;
  // Set-up event forwarding
  var forwardBlur = (0, _handle.forward)('onBlur');
  var forwardChange = (0, _handle.forward)('onChange');
  var forwardClick = (0, _handle.forward)('onClick');
  var forwardKeyDown = (0, _handle.forward)('onKeyDown');
  return _class = /*#__PURE__*/function (_Component) {
    _inherits(_class, _Component);
    var _super = _createSuper(_class);
    function _class(props) {
      var _this;
      _classCallCheck(this, _class);
      _this = _super.call(this, props);
      _this.handleChange = function (ev) {
        if (!_this.state.isActive) {
          _this.setState({
            value: _this.validateValue(parseInt(ev.value)),
            noAnimation: false
          });
          forwardChange(ev, _this.props);
        }
      };
      _this.prepareInput = function () {
        _this.setState({
          isActive: true,
          noAnimation: true
        });
        _this.freezeSpotlight(true);
      };
      _this.handleClick = function (ev) {
        if (!_this.props.disabled && ev.type === 'click') {
          _this.prepareInput();
        }
        forwardClick(ev, _this.props);
      };
      _this.getInputNode = function (node) {
        if (node) {
          _this.inputNode = node;
          _this.inputNode.focus();
        }
      };
      _this.getPickerNode = function (node) {
        _this.pickerNode = _reactDom["default"].findDOMNode(node); // eslint-disable-line react/no-find-dom-node
      };
      _this.freezeSpotlight = function (freeze) {
        if (!freeze) {
          _this.paused.resume();
        } else {
          _this.paused.pause();
        }
      };
      _this.handleKeyDown = function (ev) {
        var keyCode = ev.keyCode;
        if (_this.props.disabled) return;
        if ((0, _keymap.is)('enter', keyCode) && !_this.state.isActive || (0, _keymap.is)('numSet', keyCode) && !_this.state.isActive || (0, _keymap.is)('plus', keyCode) || (0, _keymap.is)('minus', keyCode)) {
          _this.prepareInput();
        } else if (_this.state.isActive && ((0, _keymap.is)('enter', keyCode) || (0, _keymap.is)('cancel', keyCode) || (0, _keymap.is)('up', keyCode) || (0, _keymap.is)('down', keyCode))) {
          _this.handleBlur(ev);
        }
        forwardKeyDown(ev, _this.props);
      };
      _this.handleBlur = function (ev) {
        var value = _this.validateValue(parseInt(ev.target.value));
        if (_this.state.value !== value) {
          forwardChange({
            value: value
          }, _this.props);
        }
        _this.setState({
          value: value,
          isActive: false
        }, function () {
          _spotlight["default"].focus(_this.pickerNode);
          _this.freezeSpotlight(false);
        });
        forwardBlur(ev, _this.props);
      };
      _this.validateValue = function (value) {
        return value <= _this.props.max && value >= _this.props.min ? value : _this.state.value;
      };
      _this.paused = new _Pause["default"]('EditableIntegerPickerDecorator');
      _this.state = {
        isActive: false,
        value: props.value,
        // Animation is disabled on the picker when activating edit mode and enabled on the
        // next onChange event. The `noAnimation` prop takes precedence over this.
        noAnimation: false
      };
      return _this;
    }
    _createClass(_class, [{
      key: "render",
      value: function render() {
        return /*#__PURE__*/(0, _jsxRuntime.jsx)(Wrapped, _objectSpread(_objectSpread({}, this.props), {}, {
          editMode: this.state.isActive,
          inputRef: this.getInputNode,
          joined: true,
          noAnimation: this.props.noAnimation || this.state.noAnimation,
          onChange: this.handleChange,
          onPickerItemClick: this.handleClick,
          onInputBlur: this.handleBlur,
          onKeyDown: this.handleKeyDown,
          pickerRef: this.getPickerNode,
          value: this.state.value
        }));
      }
    }]);
    return _class;
  }(_react.Component), _class.displayName = 'EditableIntegerPickerDecorator', _class.propTypes = /** @lends moonstone/EditableIntegerPicker.EditableIntegerPickerDecorator.prototype */{
    /**
    	* The maximum value selectable by the picker (inclusive).
    	*
    	* @type {Number}
    * @required
    	* @public
    	*/
    max: _propTypes["default"].number.isRequired,
    /**
     * The minimum value selectable by the picker (inclusive).
     *
     * @type {Number}
     * @required
     * @public
     */
    min: _propTypes["default"].number.isRequired,
    /**
     * Disables EditableIntegerPicker and does not generate `onChange`
     * {@link /docs/developer-guide/glossary/#event|events}.
     *
     * @type {Boolean}
     * @public
     */
    disabled: _propTypes["default"].bool,
    /**
     * Disables transition animation.
     *
     * @type {Boolean}
     * @public
     */
    noAnimation: _propTypes["default"].bool,
    /**
     * The value displayed in the picker.
     *
     * @type {Number}
     * @default 0
     * @public
     */
    value: _propTypes["default"].number
  }, _class.defaultProps = {
    value: 0
  }, _class;
});
var _default = exports["default"] = EditableIntegerPickerDecorator;
