// @enact/moonstone/DaySelector/DaySelectorCheckbox.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.DaySelectorCheckboxBase = exports.DaySelectorCheckbox = void 0;
var _kind = _interopRequireDefault(require("@enact/core/kind"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _Icon = _interopRequireDefault(require("../Icon"));
var _Skinnable = _interopRequireDefault(require("../Skinnable"));
var _DaySelectorCheckboxModule = _interopRequireDefault(require("./DaySelectorCheckbox.module.css"));
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : String(i); }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _objectDestructuringEmpty(obj) { if (obj == null) throw new TypeError("Cannot destructure " + obj); } /**
 * A component that represents the selected state of a day within a
 * {@link moonstone/DaySelector.DaySelector}. It has built-in spotlight support and is intended for
 * use in a specialized {@link moonstone/Item|Item} that does not visually respond to focus, so
 * this can show focus instead.
 *
 * @class DaySelectorCheckbox
 * @memberof moonstone/DaySelector
 * @ui
 * @private
 */
var DaySelectorCheckboxBase = exports.DaySelectorCheckboxBase = (0, _kind["default"])({
  name: 'DaySelectorCheckbox',
  propTypes: /** @lends moonstone/DaySelector.DaySelectorCheckbox.prototype */{
    /**
     * Sets whether this control is disabled, and non-interactive
     *
     * @type {Boolean}
     * @public
     */
    disabled: _propTypes["default"].bool,
    /**
     * Sets whether this control is in the "on" or "off" state. `true` for on, `false` for "off".
     *
     * @type {Boolean}
     * @public
     */
    selected: _propTypes["default"].bool
  },
  styles: {
    css: _DaySelectorCheckboxModule["default"],
    className: 'daySelectorCheckbox'
  },
  computed: {
    className: function className(_ref) {
      var selected = _ref.selected,
        styler = _ref.styler;
      return styler.append({
        selected: selected
      });
    }
  },
  render: function render(_ref2) {
    var rest = Object.assign({}, (_objectDestructuringEmpty(_ref2), _ref2));
    delete rest.selected;
    return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", _objectSpread(_objectSpread({}, rest), {}, {
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Icon["default"], {
        className: _DaySelectorCheckboxModule["default"].icon,
        children: "check"
      })
    }));
  }
});
var DaySelectorCheckbox = exports.DaySelectorCheckbox = (0, _Skinnable["default"])(DaySelectorCheckboxBase);
var _default = exports["default"] = DaySelectorCheckbox;
