// @enact/spotlight/src/inputType.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setInputType = exports.getInputType = exports.getInputInfo = exports.applyInputTypeToNode = exports.activateInputType = void 0;
/**
 * Exports methods for working with inputType in spotlight
 *
 * @module spotlight/inputType
 * @private
 */

var input = {
  activated: false,
  applied: false,
  types: {
    key: true,
    mouse: false,
    touch: false
  }
};
var activateInputType = exports.activateInputType = function activateInputType(activated) {
  input.activated = activated;
};
var getInputType = exports.getInputType = function getInputType() {
  return Object.keys(input.types).find(function (type) {
    return input.types[type];
  });
};
var setInputType = exports.setInputType = function setInputType(inputType) {
  if (Object.prototype.hasOwnProperty.call(input.types, inputType) && !input.types[inputType]) {
    Object.keys(input.types).map(function (type) {
      input.types[type] = false;
    });
    input.types[inputType] = true;
    input.applied = false;
  }
};
var applyInputTypeToNode = exports.applyInputTypeToNode = function applyInputTypeToNode(node) {
  Object.keys(input.types).map(function (type) {
    node.classList.toggle('spotlight-input-' + type, input.types[type]);
  });
  input.applied = true;
};
var getInputInfo = exports.getInputInfo = function getInputInfo() {
  return {
    activated: input.activated,
    applied: input.applied
  };
};
