// @enact/moonstone/internal/Picker/PickerButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.PickerButtonBase = exports.PickerButton = void 0;
var _handle = require("@enact/core/handle");
var _kind = _interopRequireDefault(require("@enact/core/kind"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _MarqueeController = require("@enact/ui/Marquee/MarqueeController");
var _Pure = _interopRequireDefault(require("@enact/ui/internal/Pure"));
var _Touchable = _interopRequireDefault(require("@enact/ui/Touchable"));
var _Icon = _interopRequireDefault(require("../../Icon"));
var _IconButton = _interopRequireDefault(require("../../IconButton"));
var _PickerModule = _interopRequireDefault(require("./Picker.module.css"));
var _jsxRuntime = require("react/jsx-runtime");
var _excluded = ["disabled", "icon"],
  _excluded2 = ["disabled", "icon", "joined"];
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : String(i); }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
var JoinedPickerButtonBase = (0, _kind["default"])({
  name: 'JoinedPickerButtonBase',
  propTypes: {
    disabled: _propTypes["default"].bool,
    icon: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].object])
  },
  render: function render(_ref) {
    var disabled = _ref.disabled,
      icon = _ref.icon,
      rest = _objectWithoutProperties(_ref, _excluded);
    return /*#__PURE__*/(0, _jsxRuntime.jsx)("span", _objectSpread(_objectSpread({}, rest), {}, {
      "data-webos-voice-intent": "Select",
      disabled: disabled,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Icon["default"], {
        className: _PickerModule["default"].icon,
        disabled: disabled,
        size: "small",
        children: icon
      })
    }));
  }
});
var JoinedPickerButton = (0, _Touchable["default"])(JoinedPickerButtonBase);
var PickerButtonBase = exports.PickerButtonBase = (0, _kind["default"])({
  name: 'PickerButton',
  propTypes: {
    disabled: _propTypes["default"].bool,
    hidden: _propTypes["default"].bool,
    icon: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].object]),
    joined: _propTypes["default"].bool,
    onSpotlightDisappear: _propTypes["default"].func,
    spotlightDisabled: _propTypes["default"].bool
  },
  styles: {
    css: _PickerModule["default"]
  },
  handlers: {
    onMouseEnter: (0, _handle.handle)((0, _handle.forward)('onMouseEnter'), function (ev, props, sync) {
      if (sync && sync.enter) {
        sync.enter(null);
      }
    }),
    onMouseLeave: (0, _handle.handle)((0, _handle.forward)('onMouseLeave'), function (ev, props, sync) {
      if (sync && sync.leave) {
        sync.leave(null);
      }
    })
  },
  computed: {
    className: function className(_ref2) {
      var hidden = _ref2.hidden,
        styler = _ref2.styler;
      return styler.append({
        hidden: hidden
      });
    }
  },
  render: function render(_ref3) {
    var disabled = _ref3.disabled,
      icon = _ref3.icon,
      joined = _ref3.joined,
      rest = _objectWithoutProperties(_ref3, _excluded2);
    delete rest.hidden;
    if (joined) {
      delete rest.onSpotlightDisappear;
      delete rest.spotlightDisabled;
      return /*#__PURE__*/(0, _jsxRuntime.jsx)(JoinedPickerButton, _objectSpread(_objectSpread({}, rest), {}, {
        icon: icon,
        disabled: disabled
      }));
    } else {
      return /*#__PURE__*/(0, _jsxRuntime.jsx)(_IconButton["default"], _objectSpread(_objectSpread({}, rest), {}, {
        backgroundOpacity: "transparent",
        disabled: disabled,
        size: "small",
        children: icon
      }));
    }
  }
});

// This can be replaced with the kind config contextType when it's supported
PickerButtonBase.contextType = _MarqueeController.MarqueeControllerContext;
var PickerButton = exports.PickerButton = (0, _Pure["default"])(PickerButtonBase);
var _default = exports["default"] = PickerButton;
