// @enact/moonstone/SelectableItem/SelectableItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.SelectableItemDecorator = exports.SelectableItemBase = exports.SelectableItem = void 0;
var _kind = _interopRequireDefault(require("@enact/core/kind"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _ToggleItem = require("../ToggleItem");
var _SelectableIcon = _interopRequireDefault(require("./SelectableIcon"));
var _SelectableItemModule = _interopRequireDefault(require("./SelectableItem.module.css"));
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : String(i); }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); } /**
 * Provides a Moonstone-themed {@link moonstone/Item|Item} with an icon that toggles on and off.
 *
 * @example
 * <SelectableItem>Click Me</SelectableItem>
 *
 * @module moonstone/SelectableItem
 * @exports SelectableItem
 * @exports SelectableItemBase
 * @exports SelectableItemDecorator
 */ /**
 * Renders an {@link moonstone/Item|Item} with a circle icon, by default.
 *
 * @class SelectableItemBase
 * @memberof moonstone/SelectableItem
 * @extends moonstone/ToggleItem.ToggleItemBase
 * @omit iconComponent
 * @ui
 * @public
 */
var SelectableItemBase = exports.SelectableItemBase = (0, _kind["default"])({
  name: 'SelectableItem',
  propTypes: /** @lends moonstone/SelectableItem.SelectableItemBase.prototype */{
    /**
     * Customizes the component by mapping the supplied collection of CSS class names to the
     * corresponding internal elements and states of this component.
     *
     * The following classes are supported:
     *
     * * `selectableItem` - The root class name
     *
     * @type {Object}
     * @public
     */
    css: _propTypes["default"].object
  },
  styles: {
    css: _SelectableItemModule["default"],
    className: 'selectableItem',
    publicClassNames: ['selectableItem']
  },
  render: function render(props) {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_ToggleItem.ToggleItemBase, _objectSpread(_objectSpread({
      "data-webos-voice-intent": "SelectCheckItem"
    }, props), {}, {
      css: props.css,
      iconComponent: _SelectableIcon["default"]
    }));
  }
});

/**
 * Adds interactive toggle functionality to `SelectableItemBase`.
 *
 * @class SelectableItemDecorator
 * @memberof moonstone/SelectableItem
 * @mixes moonstone/ToggleItem.ToggleItemDecorator
 * @hoc
 * @public
 */
var SelectableItemDecorator = exports.SelectableItemDecorator = (0, _ToggleItem.ToggleItemDecorator)({
  invalidateProps: ['inline', 'selected']
});

/**
 * A Moonstone-styled item with a toggle icon, marqueed text, and `Spotlight` focus.
 *
 * @class SelectableItem
 * @memberof moonstone/SelectableItem
 * @extends moonstone/SelectableItem.SelectableItemBase
 * @mixes moonstone/SelectableItem.SelectableItemDecorator
 * @ui
 * @public
 */
var SelectableItem = exports.SelectableItem = SelectableItemDecorator(SelectableItemBase);
var _default = exports["default"] = SelectableItem;
