// @enact/moonstone/SwitchItem/SwitchItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.SwitchItemBase = exports.SwitchItem = void 0;
var _kind = _interopRequireDefault(require("@enact/core/kind"));
var _ComponentOverride = _interopRequireDefault(require("@enact/ui/ComponentOverride"));
var _propTypes = _interopRequireDefault(require("@enact/core/internal/prop-types"));
var _propTypes2 = _interopRequireDefault(require("prop-types"));
var _Switch = _interopRequireDefault(require("../Switch"));
var _ToggleItem = _interopRequireDefault(require("../ToggleItem"));
var _SwitchItemModule = _interopRequireDefault(require("./SwitchItem.module.css"));
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : String(i); }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); } /**
 * Provides Moonstone-themed item component and interactive toggleable switch.
 *
 * @example
 * <SwitchItem>
 * 	Item
 * </SwitchItem>
 *
 * @module moonstone/SwitchItem
 * @exports SwitchItem
 * @exports SwitchItemBase
 */ /**
 * Renders an item with a {@link moonstone/Switch|Switch}.
 *
 * @class SwitchItem
 * @memberof moonstone/SwitchItem
 * @extends moonstone/ToggleItem.ToggleItem
 * @omit iconComponent
 * @ui
 * @public
 */
var SwitchItemBase = exports.SwitchItemBase = exports.SwitchItem = (0, _kind["default"])({
  name: 'SwitchItem',
  propTypes: /** @lends moonstone/SwitchItem.SwitchItem.prototype */{
    /**
     * Customizes the component by mapping the supplied collection of CSS class names to the
     * corresponding internal elements and states of this component.
     *
     * The following classes are supported:
     *
     * * `switchItem` - The root class name
     *
     * @type {Object}
     * @public
     */
    css: _propTypes2["default"].object,
    /**
     * Customize the component used as the switch.
     *
     * @type {Element|Component}
     * @default {@link moonstone/Switch.Switch}
     * @private
     */
    iconComponent: _propTypes["default"].componentOverride
  },
  defaultProps: {
    iconComponent: _Switch["default"]
  },
  styles: {
    css: _SwitchItemModule["default"],
    className: 'switchItem',
    publicClassNames: ['switchItem']
  },
  computed: {
    iconComponent: function iconComponent(_ref) {
      var css = _ref.css,
        _iconComponent = _ref.iconComponent;
      return /*#__PURE__*/(0, _jsxRuntime.jsx)(_ComponentOverride["default"], {
        component: _iconComponent,
        className: css["switch"]
      });
    }
  },
  render: function render(props) {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_ToggleItem["default"], _objectSpread(_objectSpread({
      "data-webos-voice-intent": "SetToggleItem"
    }, props), {}, {
      css: props.css,
      iconPosition: "after"
    }));
  }
});
var _default = exports["default"] = SwitchItemBase;
