// @enact/ui/Measurable/Measurable.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useMeasurable = exports["default"] = exports.Measurable = void 0;
var _hoc = _interopRequireDefault(require("@enact/core/hoc"));
var _react = require("react");
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : String(i); }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t["return"] && (u = t["return"](), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; } /* global ResizeObserver, MutationObserver */ /**
 * A higher-order component that adds the ability to measure nodes conveniently.
 *
 * @module ui/Measurable
 * @exports Measurable
 */ /**
 * Default config for {@link ui/Measurable.Measurable}.
 *
 * @memberof ui/Measurable.Measurable
 * @hocconfig
 */
var defaultConfig = {
  /**
   * Configures the prop name to pass the measurement object.
   *
   * @type {String}
   * @default 'measurement'
   * @memberof ui/Measurable.Measurable.defaultConfig
   */
  measurementProp: 'measurement',
  /**
   * Configures the prop name to pass a function to capture a reference to the element to measure.
   *
   * @type {String}
   * @default 'forwardRef'
   * @memberof ui/Measurable.Measurable.defaultConfig
   */
  refProp: 'forwardRef'
};
var useMeasurable = exports.useMeasurable = function useMeasurable() {
  var _useState = (0, _react.useState)(),
    _useState2 = _slicedToArray(_useState, 2),
    measurement = _useState2[0],
    setMeasurement = _useState2[1];
  var ref = (0, _react.useRef)(null);
  (0, _react.useLayoutEffect)(function () {
    var refToMeasure = ref.current;
    if (!refToMeasure) {
      return;
    }

    // Support for ResizeObserver
    if (typeof ResizeObserver === 'function') {
      var resizeObserver = new ResizeObserver(function (entries) {
        entries.forEach(function (entry) {
          if (entry.target === refToMeasure) {
            // we want to measure including the padding, hence refers to `target` instead of `contentRect`
            setMeasurement(entry.target.getBoundingClientRect());
          }
        });
      });
      resizeObserver.observe(refToMeasure);
      return function () {
        resizeObserver.disconnect(refToMeasure);
        resizeObserver = null;
      };
      // Fallback support for MutationObserver
    } else if (typeof MutationObserver === 'function') {
      var mutationObserver = new MutationObserver(function (mutationsList) {
        var _iterator = _createForOfIteratorHelper(mutationsList),
          _step;
        try {
          for (_iterator.s(); !(_step = _iterator.n()).done;) {
            var mutation = _step.value;
            if (mutation.type === 'childList') {
              setMeasurement(mutation.target.getBoundingClientRect());
            } else if (mutation.type === 'attributes' && (mutation.attributeName === 'style' || mutation.attributeName === 'class')) {
              setMeasurement(refToMeasure.getBoundingClientRect());
            }
          }
        } catch (err) {
          _iterator.e(err);
        } finally {
          _iterator.f();
        }
      });
      mutationObserver.observe(refToMeasure, {
        attributes: true,
        childList: true,
        subtree: true
      });
      return function () {
        mutationObserver.disconnect();
        mutationObserver = null;
      };
    }
  }, [] // disconnect on unmount
  );
  return {
    ref: ref,
    measurement: measurement
  };
};

/**
 * A higher-order component that adds the ability to measure a referenced node and get that value
 * back as a prop.
 *
 * Applying `Measurable` to a component will pass two additional props: function to fetch the `ref`
 * and the measurement object (currently this executes `getBoundingClientRect` returning its full
 * complement of properties). Both prop names are configurable through the HOC config object.
 *
 * Changes to the referenced node will result in new measurements being returned as props. That
 * being said, it probably is not wise to use this during animation as it may cause unwanted
 * performance issues.
 *
 * @see https://reactjs.org/docs/hooks-faq.html#how-can-i-measure-a-dom-node
 * @class Measurable
 * @memberof ui/Measurable
 * @hoc
 * @public
 */
var Measurable = exports.Measurable = (0, _hoc["default"])(defaultConfig, function (configHoc, Wrapped) {
  return function MeasurableHoc(props) {
    // Take the config from Measurable and insert it into a fresh instance of defineMeasurable
    var _useMeasurable = useMeasurable(),
      ref = _useMeasurable.ref,
      measurement = _useMeasurable.measurement;
    var measurementProps = _defineProperty(_defineProperty({}, configHoc.refProp, ref), configHoc.measurementProp, measurement);
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(Wrapped, _objectSpread(_objectSpread({}, props), measurementProps));
  };
});
var _default = exports["default"] = Measurable;
