// @enact/sandstone/FormCheckboxItem/FormCheckboxItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.FormCheckboxItemDecorator = exports.FormCheckboxItemBase = exports.FormCheckboxItem = void 0;
var _kind = _interopRequireDefault(require("@enact/core/kind"));
var _Pure = _interopRequireDefault(require("@enact/ui/internal/Pure"));
var _Slottable = _interopRequireDefault(require("@enact/ui/Slottable"));
var _Toggleable = _interopRequireDefault(require("@enact/ui/Toggleable"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _compose = _interopRequireDefault(require("ramda/src/compose"));
var _react = require("react");
var _Skinnable = _interopRequireDefault(require("../Skinnable"));
var _Checkbox = require("../Checkbox");
var _Item = require("../Item");
var _FormCheckboxItemModule = _interopRequireDefault(require("./FormCheckboxItem.module.css"));
var _jsxRuntime = require("react/jsx-runtime");
var _excluded = ["children", "css", "icon", "indeterminate", "indeterminateIcon", "selected", "slotBefore"];
/**
 * Provides Sandstone styled form item component and interactive toggleable checkbox.
 *
 * @example
 * <FormCheckboxItem>A Checkbox for a form</FormCheckboxItem>
 *
 * @module sandstone/FormCheckboxItem
 * @exports FormCheckboxItem
 * @exports FormCheckboxItemBase
 * @exports FormCheckboxItemDecorator
 */
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
var Item = (0, _Item.ItemDecorator)(_Item.ItemBase);
var Checkbox = (0, _Skinnable["default"])(_Checkbox.CheckboxBase);
var hasChildren = function hasChildren(children) {
  return _react.Children.toArray(children).filter(Boolean).length > 0;
};

/**
 * A Sandstone-styled form item with a checkbox component.
 *
 * Useful to show a selected state on an item inside a form.
 *
 * @class FormCheckboxItemBase
 * @memberof sandstone/FormCheckboxItem
 * @extends sandstone/Item.Item
 * @ui
 * @public
 */
var FormCheckboxItemBase = exports.FormCheckboxItemBase = (0, _kind["default"])({
  name: 'FormCheckboxItem',
  propTypes: /** @lends sandstone/FormCheckboxItem.FormCheckboxItemBase.prototype */{
    /**
     * Customizes the component by mapping the supplied collection of CSS class names to the
     * corresponding internal elements and states of this component.
     *
     * The following classes are supported:
     *
     * * `formCheckboxItem` - The root class name
     *
     * @type {Object}
     * @public
     */
    css: _propTypes["default"].object,
    /**
     * The icon content.
     *
     * May be specified as either:
     *
     * * A string that represents an icon from the {@link sandstone/Icon.iconList|iconList},
     * * An HTML entity string, Unicode reference or hex value (in the form '0x...'),
     * * A URL specifying path to an icon image, or
     * * An object representing a resolution independent resource (See {@link ui/resolution})
     *
     * @type {String|Object}
     * @public
     */
    icon: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].object]),
    /**
     * Enables the "indeterminate" state.
     *
     * An indeterminate, mixed, or half-selected state is typically used in a hierarchy or group
     * to represent that some, not all, children are selected.
     *
     * NOTE: This does not prevent updating the `selected` state. Applications must control this
     * property directly.
     *
     * @type {Boolean}
     * @public
     */
    indeterminate: _propTypes["default"].bool,
    /**
     * The icon to be used in the `indeterminate` state.
     *
     * May be specified as either:
     *
     * * A string that represents an icon from the {@link sandstone/Icon.iconList|iconList},
     * * An HTML entity string, Unicode reference or hex value (in the form '0x...'),
     * * A URL specifying path to an icon image, or
     * * An object representing a resolution independent resource (See {@link ui/resolution})
     *
     * @type {String}
     * @public
     */
    indeterminateIcon: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].object]),
    /**
     * Controls the presence of the checkmark icon.
     *
     * @type {Boolean}
     * @public
     */
    selected: _propTypes["default"].bool,
    /**
     * Nodes to be inserted after the checkbox and before `children`.
     *
     * @type {Node}
     * @public
     */
    slotBefore: _propTypes["default"].node
  },
  styles: {
    css: _FormCheckboxItemModule["default"],
    className: 'formCheckboxItem',
    publicClassNames: ['formCheckboxItem']
  },
  computed: {
    className: function className(_ref) {
      var slotBefore = _ref.slotBefore,
        styler = _ref.styler;
      return styler.append({
        hasSlotBefore: hasChildren(slotBefore)
      });
    }
  },
  render: function render(_ref2) {
    var children = _ref2.children,
      css = _ref2.css,
      icon = _ref2.icon,
      indeterminate = _ref2.indeterminate,
      indeterminateIcon = _ref2.indeterminateIcon,
      selected = _ref2.selected,
      slotBefore = _ref2.slotBefore,
      rest = _objectWithoutProperties(_ref2, _excluded);
    return /*#__PURE__*/(0, _jsxRuntime.jsxs)(Item, _objectSpread(_objectSpread({
      "data-webos-voice-intent": "SelectCheckItem",
      role: "checkbox"
    }, rest), {}, {
      "aria-checked": selected,
      css: css,
      selected: selected,
      children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)("slotBefore", {
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(Checkbox, {
          className: css.checkbox,
          indeterminate: indeterminate,
          indeterminateIcon: indeterminateIcon,
          selected: selected,
          standalone: true,
          children: icon
        }), slotBefore]
      }), children]
    }));
  }
});

/**
 * Adds interactive functionality to `FormCheckboxItem`.
 *
 * @class FormCheckboxItemDecorator
 * @memberof sandstone/FormCheckboxItem
 * @mixes ui/Toggleable.Toggleable
 * @hoc
 * @public
 */
var FormCheckboxItemDecorator = exports.FormCheckboxItemDecorator = (0, _compose["default"])((0, _Toggleable["default"])({
  toggleProp: 'onClick'
}), (0, _Slottable["default"])({
  slots: ['label', 'slotAfter', 'slotBefore']
}));

/**
 * A Sandstone-styled form item with a checkbox component.
 *
 * `FormCheckboxItem` will manage its `selected` state via {@link ui/Toggleable|Toggleable} unless
 * set directly.
 *
 * @class FormCheckboxItem
 * @memberof sandstone/FormCheckboxItem
 * @extends sandstone/FormCheckboxItem.FormCheckboxItemBase
 * @mixes sandstone/FormCheckboxItem.FormCheckboxItemDecorator
 * @ui
 * @public
 */
var FormCheckboxItem = exports.FormCheckboxItem = (0, _Pure["default"])(FormCheckboxItemDecorator(FormCheckboxItemBase));
var _default = exports["default"] = FormCheckboxItem;
