// @enact/moonstone/ToggleButton/ToggleButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.ToggleButtonBase = exports.ToggleButton = void 0;
var _kind = _interopRequireDefault(require("@enact/core/kind"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _Pure = _interopRequireDefault(require("@enact/ui/internal/Pure"));
var _Toggleable = _interopRequireDefault(require("@enact/ui/Toggleable"));
var _Button = _interopRequireDefault(require("../Button"));
var _Skinnable = _interopRequireDefault(require("../Skinnable"));
var _ToggleButtonModule = _interopRequireDefault(require("./ToggleButton.module.css"));
var _jsxRuntime = require("react/jsx-runtime");
var _excluded = ["selected"];
/**
 * Provides a Moonstone-themed toggle button component and behaviors.
 *
 * @example
 * <ToggleButton>Toggle me</ToggleButton>
 *
 * @module moonstone/ToggleButton
 * @exports ToggleButton
 * @exports ToggleButtonBase
 */
/**
 * A stateless {@link moonstone/Button.Button|Button} that can be toggled by changing its
 * `selected` property.
 *
 * @class ToggleButtonBase
 * @memberof moonstone/ToggleButton
 * @extends moonstone/Button.Button
 * @ui
 * @public
 */
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : String(i); }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
var ToggleButtonBase = exports.ToggleButtonBase = (0, _kind["default"])({
  name: 'ToggleButton',
  propTypes: /** @lends moonstone/ToggleButton.ToggleButtonBase.prototype */{
    /**
     * The background-color opacity of this button.
     *
     * * Values: `'translucent'`, `'lightTranslucent'`, `'transparent'`
     *
     * @type {String}
     * @public
     */
    backgroundOpacity: _propTypes["default"].oneOf(['translucent', 'lightTranslucent', 'transparent']),
    /**
     * The string to be displayed as the main content of the toggle button.
     *
     * If `toggleOffLabel` and/or `toggleOnLabel` are provided, they will be used for the
     * respective states.
     *
     * @type {Node}
     * @public
     */
    children: _propTypes["default"].node,
    /**
     * Disables the button.
     *
     * @type {Boolean}
     * @default false
     * @public
     */
    disabled: _propTypes["default"].bool,
    /**
     * Enforces a minimum width on the Button.
     *
     * *NOTE*: This property's default is `true` and must be explicitly set to `false` to allow
     * the button to shrink to fit its contents.
     *
     * @type {Boolean}
     * @default true
     * @public
     */
    minWidth: _propTypes["default"].bool,
    /**
     * Applies a pressed visual effect.
     *
     * @type {Boolean}
     * @public
     */
    pressed: _propTypes["default"].bool,
    /**
     * Indicates the button is 'on'.
     *
     * @type {Boolean}
     * @default false
     * @public
     */
    selected: _propTypes["default"].bool,
    /**
     * The size of the button.
     *
     * A `'small'` button will have a larger tap target than its apparent size to allow it to be
     * clicked more easily.
     *
     * @type {('small'|'large')}
     * @default 'large'
     * @public
     */
    size: _propTypes["default"].string,
    /**
     * Button text displayed in the 'off' state.
     *
     * If not specified, `children` will be used for 'off' button text.
     *
     * @type {String}
     * @default ''
     * @public
     */
    toggleOffLabel: _propTypes["default"].string,
    /**
     * Button text displayed in the 'on' state.
     *
     * If not specified, `children` will be used for 'on' button text.
     *
     * @type {String}
     * @default ''
     * @public
     */
    toggleOnLabel: _propTypes["default"].string
  },
  defaultProps: {
    disabled: false,
    minWidth: true,
    selected: false,
    toggleOffLabel: '',
    toggleOnLabel: ''
  },
  styles: {
    css: _ToggleButtonModule["default"],
    className: 'toggleButton'
  },
  computed: {
    className: function className(_ref) {
      var selected = _ref.selected,
        size = _ref.size,
        styler = _ref.styler;
      return styler.append({
        selected: selected
      }, size);
    },
    children: function children(_ref2) {
      var _children = _ref2.children,
        selected = _ref2.selected,
        toggleOnLabel = _ref2.toggleOnLabel,
        toggleOffLabel = _ref2.toggleOffLabel;
      var c = _children;
      if (selected && toggleOnLabel) {
        c = toggleOnLabel;
      } else if (!selected && toggleOffLabel) {
        c = toggleOffLabel;
      }
      return c;
    }
  },
  render: function render(_ref3) {
    var selected = _ref3.selected,
      rest = _objectWithoutProperties(_ref3, _excluded);
    delete rest.toggleOffLabel;
    delete rest.toggleOnLabel;
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_Button["default"], _objectSpread(_objectSpread({
      "data-webos-voice-intent": "SetToggleItem"
    }, rest), {}, {
      "aria-pressed": selected,
      css: _ToggleButtonModule["default"],
      decoration: /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
        className: _ToggleButtonModule["default"].toggleIndicator
      }),
      selected: selected
    }));
  }
});

/**
 * A toggleable button.
 *
 * By default, `ToggleButton` maintains the state of its `selected` property.
 * Supply the `defaultSelected` property to control its initial value. If you
 * wish to directly control updates to the component, supply a value to `selected` at creation time
 * and update it in response to `onToggle` events.
 *
 * @class ToggleButton
 * @memberof moonstone/ToggleButton
 * @extends moonstone/ToggleButton.ToggleButtonBase
 * @ui
 * @mixes ui/Toggleable.Toggleable
 * @public
 */
var ToggleButton = exports.ToggleButton = (0, _Pure["default"])((0, _Toggleable["default"])({
  prop: 'selected',
  toggleProp: 'onTap'
}, (0, _Skinnable["default"])(ToggleButtonBase)));
var _default = exports["default"] = ToggleButton;
