// @enact/moonstone/MoonstoneDecorator/MoonstoneDecorator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.MoonstoneDecorator = void 0;
var _dispatcher = require("@enact/core/dispatcher");
var _keymap = require("@enact/core/keymap");
var _hoc = _interopRequireDefault(require("@enact/core/hoc"));
var _I18nDecorator = _interopRequireDefault(require("@enact/i18n/I18nDecorator"));
var _react = require("react");
var _classnames = _interopRequireDefault(require("classnames"));
var _resolution = require("@enact/ui/resolution");
var _FloatingLayer = require("@enact/ui/FloatingLayer");
var _SpotlightRootDecorator = _interopRequireDefault(require("@enact/spotlight/SpotlightRootDecorator"));
var _Skinnable = _interopRequireDefault(require("../Skinnable"));
var _I18nFontDecorator = _interopRequireDefault(require("./I18nFontDecorator"));
var _AccessibilityDecorator = _interopRequireDefault(require("./AccessibilityDecorator"));
var _screenTypes = _interopRequireDefault(require("./screenTypes.json"));
var _MoonstoneDecoratorModule = _interopRequireDefault(require("./MoonstoneDecorator.module.css"));
var _Touchable = require("@enact/ui/Touchable");
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); Object.defineProperty(subClass, "prototype", { writable: false }); if (superClass) _setPrototypeOf(subClass, superClass); }
function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
function _possibleConstructorReturn(self, call) { if (call && (typeof call === "object" || typeof call === "function")) { return call; } else if (call !== void 0) { throw new TypeError("Derived constructors may only return object or undefined"); } return _assertThisInitialized(self); }
function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : String(i); }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); } /**
 * Exports the {@link moonstone/MoonstoneDecorator.MoonstoneDecorator|MoonstoneDecorator} HOC
 *
 * @module moonstone/MoonstoneDecorator
 * @exports MoonstoneDecorator
 */ /**
 * Default config for `MoonstoneDecorator`.
 *
 * @memberof moonstone/MoonstoneDecorator.MoonstoneDecorator
 * @hocconfig
 */
var defaultConfig = /** @lends moonstone/MoonstoneDecorator.MoonstoneDecorator.defaultConfig */{
  /**
   * Applies AccessibilityDecorator.
   *
   * If not applied, app will not support accessibility options.
   *
   * @type {Boolean}
   * @default true
   * @see {@link moonstone/MoonstoneDecorator.AccessibilityDecorator}
   * @public
   */
  accessible: true,
  /**
   * Disables use of full screen.
   *
   * @type {Boolean}
   * @default false
   * @public
   */
  disableFullscreen: false,
  /**
   * Enables a floating layer for popup components.
   *
   * If not applied, app will be responsible for applying the decorator.
   *
   * @type {Boolean}
   * @default true
   * @see {@link ui/FloatingLayer.FloatingLayerDecorator}
   * @public
   */
  "float": true,
  /**
   * Options for I18nDecorator.
   *
   * May be `false` to prevent applying the decorator. If not applied, app will be responsible for
   * applying the decorator.
   *
   * @type {Object|false}
   * @default {sync: true}
   * @see {@link i18n/I18nDecorator}
   * @public
   */
  i18n: {
    sync: true
  },
  /**
   * Disables setting spotlight focus on first render.
   *
   * @type {Boolean}
   * @default false
   * @public
   */
  noAutoFocus: false,
  /**
   * Enables overlay mode (no background color will be applied).
   *
   * @type {Boolean}
   * @default false
   * @public
   */
  overlay: false,
  /**
   * Override the resolution independence settings.
   *
   * @type {Object}
   * @see {@link ui/resolution}
   * @public
   */
  ri: {
    screenTypes: _screenTypes["default"]
  },
  /**
   * Specifies the id of the React DOM tree root node
   *
   * @type {String}
   * @default 'root'
   * @public
   */
  rootId: 'root',
  /**
   * Applies skinning support.
   *
   * @type {Boolean}
   * @default true
   * @see {@link moonstone/Skinnable}
   * @public
   */
  skin: true,
  /**
   * Applies spotlight decorator.
   *
   * If not applied, app will be responsible for applying the decorator.
   *
   * @type {Boolean}
   * @default true
   * @see {@link spotlight/SpotlightRootDecorator}
   * @public
   */
  spotlight: true
};

/**
 * A higher-order component that applies Moonstone theming to an application.
 *
 * It also applies {@link ui/FloatingLayer.FloatingLayerDecorator|floating layer}, {@link ui/resolution.ResolutionDecorator|resolution independence},
 * {@link moonstone/Skinnable|skin support}, {@link spotlight.SpotlightRootDecorator|spotlight}, and
 * {@link i18n/I18nDecorator.I18nDecorator|internationalization support}. It is meant to be applied
 * to the root element of an app.
 *
 * {@link moonstone/Skinnable|Skins} provide a way to change the coloration of your app. The
 * currently supported skins for Moonstone are "moonstone" (the default, dark skin) and
 * "moonstone-light". Use the `skin` property to assign a skin. Ex: `<DecoratedApp skin="light" />`
 *
 * Note: This HoC passes `className` to the wrapped component. It must be passed to the main DOM
 * node.
 *
 * @class MoonstoneDecorator
 * @memberof moonstone/MoonstoneDecorator
 * @mixes ui/FloatingLayer.FloatingLayerDecorator
 * @mixes ui/resolution.ResolutionDecorator
 * @mixes spotlight/SpotlightRootDecorator.SpotlightRootDecorator
 * @mixes moonstone/Skinnable.Skinnable
 * @mixes moonstone/MoonstoneDecorator.AccessibilityDecorator
 * @hoc
 * @public
 */
var MoonstoneDecorator = exports.MoonstoneDecorator = (0, _hoc["default"])(defaultConfig, function (config, Wrapped) {
  var _class;
  var accessible = config.accessible,
    ri = config.ri,
    i18n = config.i18n,
    spotlight = config.spotlight,
    _float = config["float"],
    noAutoFocus = config.noAutoFocus,
    overlay = config.overlay,
    skin = config.skin,
    disableFullscreen = config.disableFullscreen,
    rootId = config.rootId;

  // Apply classes depending on screen type (overlay / fullscreen)
  var bgClassName = (0, _classnames["default"])(_defineProperty({
    'enact-fit': !disableFullscreen
  }, _MoonstoneDecoratorModule["default"].bg, !overlay));
  var App = Wrapped;
  if (_float) App = (0, _FloatingLayer.FloatingLayerDecorator)({
    wrappedClassName: bgClassName
  }, App);
  if (ri) App = (0, _resolution.ResolutionDecorator)(ri, App);
  if (i18n) {
    // Apply the @enact/i18n decorator around the font decorator so the latter will update the
    // font stylesheet when the locale changes
    App = (0, _I18nDecorator["default"])(_objectSpread(_objectSpread({}, i18n), {}, {
      // We use the latin fonts (with non-Latin fallback) for these languages (even though
      // their scripts are non-latin)
      latinLanguageOverrides: ['ko', 'ha', 'el', 'bg', 'mk', 'mn', 'ru', 'uk', 'kk'],
      // We use the non-latin fonts for these languages (even though their scripts are
      // technically considered latin)
      nonLatinLanguageOverrides: ['en-JP']
    }), (0, _I18nFontDecorator["default"])(App));
  }
  if (spotlight) App = (0, _SpotlightRootDecorator["default"])({
    noAutoFocus: noAutoFocus
  }, App);
  if (skin) App = (0, _Skinnable["default"])({
    defaultSkin: 'dark'
  }, App);
  if (accessible) App = (0, _AccessibilityDecorator["default"])(App);

  // add webOS-specific key maps
  (0, _keymap.addAll)({
    cancel: 461,
    nonModal: [461, 415,
    // play
    19,
    // pause
    403,
    // red
    404,
    // green
    405,
    // yellow
    406,
    // blue
    33,
    // channel up
    34 // channel down
    ],
    red: 403,
    green: 404,
    yellow: 405,
    blue: 406,
    play: 415,
    pause: 19,
    rewind: 412,
    fastForward: 417,
    pointerHide: 1537,
    pointerShow: 1536
  });

  // set the DOM node ID of the React DOM tree root
  (0, _dispatcher.setDefaultTargetById)(rootId);

  // configure the default hold time
  (0, _Touchable.configure)({
    hold: {
      events: [{
        name: 'hold',
        time: 400
      }]
    }
  });
  var Decorator = (_class = /*#__PURE__*/function (_Component) {
    _inherits(Decorator, _Component);
    var _super = _createSuper(Decorator);
    function Decorator() {
      _classCallCheck(this, Decorator);
      return _super.apply(this, arguments);
    }
    _createClass(Decorator, [{
      key: "render",
      value: function render() {
        var className = (0, _classnames["default"])(_MoonstoneDecoratorModule["default"].root, this.props.className, 'enact-unselectable', _defineProperty(_defineProperty({}, bgClassName, !_float), 'enact-fit', !disableFullscreen));
        return /*#__PURE__*/(0, _jsxRuntime.jsx)(App, _objectSpread(_objectSpread({}, this.props), {}, {
          className: className
        }));
      }
    }]);
    return Decorator;
  }(_react.Component), _class.displayName = 'MoonstoneDecorator', _class);
  return Decorator;
});
var _default = exports["default"] = MoonstoneDecorator;
