// @enact/sandstone/Heading/Heading.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.HeadingDecorator = exports.HeadingBase = exports.Heading = void 0;
var _kind = _interopRequireDefault(require("@enact/core/kind"));
var _Pure = _interopRequireDefault(require("@enact/ui/internal/Pure"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _compose = _interopRequireDefault(require("ramda/src/compose"));
var _Heading = require("@enact/ui/Heading");
var _Marquee = require("../Marquee");
var _Skinnable = _interopRequireDefault(require("../Skinnable"));
var _HeadingModule = _interopRequireDefault(require("./Heading.module.css"));
var _excluded = ["css"];
/**
 * Sandstone styled labeled Heading components and behaviors
 *
 * @example
 * <Heading
 *   size="large"
 *   spacing="small"
 * >
 *   A Content Section Heading
 * </Heading>
 *
 * @module sandstone/Heading
 * @exports Heading
 * @exports HeadingBase
 * @exports HeadingDecorator
 */
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
/**
 * A labeled Heading component.
 *
 * This component is most often not used directly but may be composed within another component as it
 * is within {@link sandstone/Heading.Heading|Heading}.
 *
 * @class HeadingBase
 * @memberof sandstone/Heading
 * @ui
 * @public
 */
var HeadingBase = exports.HeadingBase = (0, _kind["default"])({
  name: 'Heading',
  propTypes: /** @lends sandstone/Heading.HeadingBase.prototype */{
    css: _propTypes["default"].object,
    /**
     * Adds a horizontal-rule (line) under the component
     *
     * @type {Boolean}
     * @public
     */
    showLine: _propTypes["default"].bool,
    /**
     * The size for slotBefore and slotAfter.
     * This size is used for remeasuring marquee metrics for Panels.Header.
     *
     * @type {String}
     * @private
     */
    slotSize: _propTypes["default"].string,
    /**
     * The size of the spacing around the Heading.
     *
     * Allowed values include:
     * * `'auto'` - Value is based on the `size` prop for automatic usage.
     * * `'large'` - Specifically assign the `'large'` spacing.
     * * `'medium'` - Specifically assign the `'medium'` spacing.
     * * `'small'` - Specifically assign the `'small'` spacing.
     * * `'none'` - No spacing at all. Neighboring elements will directly touch the Heading.
     *
     * @type {('auto'|'large'|'medium'|'small'|'none')}
     * @default 'small'
     * @public
     */
    spacing: _propTypes["default"].oneOf(['auto', 'large', 'medium', 'small', 'none'])
  },
  defaultProps: {
    spacing: 'small'
  },
  styles: {
    css: _HeadingModule["default"],
    className: 'heading'
  },
  computed: {
    className: function className(_ref) {
      var showLine = _ref.showLine,
        styler = _ref.styler;
      return styler.append({
        showLine: showLine
      });
    }
  },
  render: function render(_ref2) {
    var css = _ref2.css,
      rest = _objectWithoutProperties(_ref2, _excluded);
    delete rest.showLine;
    delete rest.slotSize;
    return _Heading.HeadingBase.inline(_objectSpread({
      css: css
    }, rest));
  }
});

/**
 * Applies Sandstone specific behaviors to {@link sandstone/Heading.HeadingBase|HeadingBase}.
 *
 * @hoc
 * @memberof sandstone/Heading
 * @mixes sandstone/Marquee.MarqueeDecorator
 * @mixes sandstone/Skinnable.Skinnable
 * @public
 */
var HeadingDecorator = exports.HeadingDecorator = (0, _compose["default"])(_Pure["default"], (0, _Marquee.MarqueeDecorator)({
  invalidateProps: ['remeasure', 'slotSize']
}), _Skinnable["default"]);

/**
 * A labeled Heading component, ready to use in Sandstone applications.
 *
 * `Heading` may be used as a header to group related components.
 *
 * Usage:
 * ```
 * <Heading
 *   spacing="medium"
 * >
 *   Related Settings
 * </Heading>
 * <CheckboxItem>A Setting</CheckboxItem>
 * <CheckboxItem>A Second Setting</CheckboxItem>
 * ```
 *
 * @class Heading
 * @memberof sandstone/Heading
 * @extends sandstone/Heading.HeadingBase
 * @mixes sandstone/Heading.HeadingDecorator
 * @ui
 * @public
 */
var Heading = exports.Heading = HeadingDecorator(HeadingBase);

/**
 * Marquee animation trigger.
 *
 * Allowed values include:
 * * `'hover'` - Marquee begins when the pointer enters the component
 * * `'render'` - Marquee begins when the component is rendered
 *
 * @name marqueeOn
 * @type {('hover'|'render')}
 * @default 'render'
 * @memberof sandstone/Heading.Heading.prototype
 * @see {@link sandstone/Marquee.Marquee}
 * @public
 */

Heading.defaultProps = {
  marqueeOn: 'render'
};
var _default = exports["default"] = Heading;
