// @enact/ui/AnnounceDecorator/AnnounceDecorator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Announce", {
  enumerable: true,
  get: function get() {
    return _Announce["default"];
  }
});
exports["default"] = exports.AnnounceDecorator = void 0;
Object.defineProperty(exports, "useAnnounce", {
  enumerable: true,
  get: function get() {
    return _useAnnounce["default"];
  }
});
var _hoc = _interopRequireDefault(require("@enact/core/hoc"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _Announce = _interopRequireDefault(require("./Announce"));
var _useAnnounce = _interopRequireDefault(require("./useAnnounce"));
var _jsxRuntime = require("react/jsx-runtime");
var _excluded = ["children"];
/**
 * Provides components to assist in notifying the user through ARIA alerts.
 *
 * @module ui/AnnounceDecorator
 * @exports Announce
 * @exports AnnounceDecorator
 */
/**
 * Default config for {@link ui/AnnounceDecorator.AnnounceDecorator}.
 *
 * @memberof ui/AnnounceDecorator.AnnounceDecorator
 * @hocconfig
 * @private
 */
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : String(i); }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
var defaultConfig = {
  /**
   * Configures the prop name to pass the announce function.
   *
   * @type {String}
   * @default 'announce'
   * @memberof ui/AnnounceDecorator.AnnounceDecorator.defaultConfig
   */
  prop: 'announce'
};

/**
 * A higher-order component that provides a function to its wrapped component which can be
 * called to alert the user for accessibility notifications.
 *
 * By default, the function is passed in the `announce` prop but may be customized by specifying the
 * `prop` config member. In addition to the configured prop, this HoC also adds an additional child
 * component to the passed `children` prop.
 *
 * Example:
 * ```
 * import AnnounceDecorator from '@enact/ui/AnnounceDecorator';
 * import {Component} from 'react';
 *
 * const Example = AnnounceDecorator(class extends Component {
 *   static propTypes = {
 *     // passed by AnnounceDecorator
 *     announce: PropTypes.func
 *   }
 *
 *   notify = () => {
 *     const {announce} = this.props;
 *     announce('this text will be alerted to user by TTS');
 *   }
 *
 *   render () {
 *     <div>
 *       <button onClick={this.notify}>Notify on Click</button>
 *     </div>
 *   }
 * });
 * ```
 *
 * @class AnnounceDecorator
 * @memberof ui/AnnounceDecorator
 * @hoc
 * @public
 */
var AnnounceDecorator = exports.AnnounceDecorator = (0, _hoc["default"])(defaultConfig, function (_ref, Wrapped) {
  var prop = _ref.prop;
  // eslint-disable-next-line no-shadow
  function AnnounceDecorator(_ref2) {
    var children = _ref2.children,
      rest = _objectWithoutProperties(_ref2, _excluded);
    var announce = (0, _useAnnounce["default"])();
    if (prop) rest[prop] = announce.announce;
    return /*#__PURE__*/(0, _jsxRuntime.jsxs)(Wrapped, _objectSpread(_objectSpread({}, rest), {}, {
      children: [children, announce.children]
    }));
  }
  AnnounceDecorator.propTypes = /** @lends ui/AnnounceDecorator.AnnounceDecorator.prototype */{
    /**
     * The wrapped component's children.
     *
     * An instance of {@link ui/AnnounceDecorator.Announce} will be appended to `children`.
     *
     * @type {Node}
     * @public
     */
    children: _propTypes["default"].node
  };
  return AnnounceDecorator;
});
var _default = exports["default"] = AnnounceDecorator;
