// @enact/ui/Item/Item.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.ItemDecorator = exports.ItemBase = exports.Item = void 0;
var _propTypes = _interopRequireDefault(require("@enact/core/internal/prop-types"));
var _kind = _interopRequireDefault(require("@enact/core/kind"));
var _propTypes2 = _interopRequireDefault(require("prop-types"));
var _compose = _interopRequireDefault(require("ramda/src/compose"));
var _Touchable = _interopRequireDefault(require("../Touchable"));
var _ForwardRef = _interopRequireDefault(require("../ForwardRef"));
var _ItemModule = _interopRequireDefault(require("./Item.module.css"));
var _jsxRuntime = require("react/jsx-runtime");
var _excluded = ["component", "componentRef", "disabled", "children"];
/**
 * An unstyled item component and behaviors to be customized by a theme or application.
 *
 * @module ui/Item
 * @exports Item
 * @exports ItemBase
 * @exports ItemDecorator
 */
/**
 * A basic list item component structure without any behaviors applied to it.
 *
 * It also has support for overlay to place things before and after the main content.
 *
 * @class ItemBase
 * @memberof ui/Item
 * @ui
 * @public
 */
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : String(i); }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
var ItemBase = exports.ItemBase = (0, _kind["default"])({
  name: 'ui:Item',
  propTypes: /** @lends ui/Item.ItemBase.prototype */{
    /**
     * The node to be displayed as the main content of the item.
     *
     * @type {Node}
     * @public
     */
    children: _propTypes2["default"].node,
    /**
     * The type of component to use to render the item. May be a DOM node name (e.g 'div',
     * 'span', etc.) or a custom component.
     *
     * @type {String|Component}
     * @default 'div'
     * @public
     */
    component: _propTypes["default"].renderable,
    /**
     * Called with a reference to {@link ui/Item.ItemBase.component|component}
     *
     * @type {Object|Function}
     * @private
     */
    componentRef: _propTypes["default"].ref,
    /**
     * Customizes the component by mapping the supplied collection of CSS class names to the
     * corresponding internal elements and states of this component.
     *
     * The following classes are supported:
     *
     * * `item` - The root class name
     * * `inline` - Applied when `inline` prop is true
     *
     * @type {Object}
     * @public
     */
    css: _propTypes2["default"].object,
    /**
     * Applies a disabled state to the item.
     *
     * @type {Boolean}
     * @default false
     * @public
     */
    disabled: _propTypes2["default"].bool,
    /**
     * Applies inline styling to the item.
     *
     * @type {Boolean}
     * @default false
     * @public
     */
    inline: _propTypes2["default"].bool
  },
  defaultProps: {
    component: 'div',
    inline: false
  },
  styles: {
    css: _ItemModule["default"],
    className: 'item',
    publicClassNames: true
  },
  computed: {
    className: function className(_ref) {
      var inline = _ref.inline,
        styler = _ref.styler;
      return styler.append({
        inline: inline
      });
    }
  },
  render: function render(_ref2) {
    var Component = _ref2.component,
      componentRef = _ref2.componentRef,
      disabled = _ref2.disabled,
      children = _ref2.children,
      rest = _objectWithoutProperties(_ref2, _excluded);
    delete rest.inline;
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(Component, _objectSpread(_objectSpread({
      ref: componentRef
    }, rest), {}, {
      "aria-disabled": disabled,
      disabled: disabled,
      children: children
    }));
  }
});

/**
 * A higher-order component that adds touch support to the component it wraps.
 *
 * @hoc
 * @memberof ui/Item
 * @mixes ui/ForwardRef.ForwardRef
 * @mixes ui/Touchable.Touchable
 * @public
 */
var ItemDecorator = exports.ItemDecorator = (0, _compose["default"])((0, _ForwardRef["default"])({
  prop: 'componentRef'
}), _Touchable["default"]);

/**
 * A minimally-styled ready-to-use list item component with touch support.
 *
 * @class Item
 * @memberof ui/Item
 * @extends ui/Item.ItemBase
 * @mixes ui/Item.ItemDecorator
 * @omit componentRef
 * @ui
 * @public
 */
var Item = exports.Item = ItemDecorator(ItemBase);
var _default = exports["default"] = Item;
